### PR TITLE
ENH: Add savgol

### LINKED ISSUE #347: Savitzky-Golay filter
add support of the Savitzky-Golay filter

see : https://github.com/scipy/scipy/pull/304

and : 

http://inst.eecs.berkeley.edu/~ee123/fa12/docs/SGFilter.pdf

could be used to filter Evoked and Epochs.

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #347.

@agramfort see what you think of the API.

### CODE REVIEW

File: mne/filter.py [hunk 1]
```diff
@@ -1389,3 +1389,62 @@ def _get_filter_length(filter_length, sfreq, min_length=128, len_x=np.inf):
         if not isinstance(filter_length, integer_types):
             raise ValueError('filter_length must be str, int, or None')
     return filter_length
+
+
+class FilterMixin(object):
+    """Object for Epoch/Evoked filtering"""
+
+    def filter(self, method='savgol', h_freq=None):
```
Comment (agramfort) [inline mne/filter.py]: why h_freq None?

File: mne/filter.py [hunk 2]
```diff
@@ -1389,3 +1389,62 @@ def _get_filter_length(filter_length, sfreq, min_length=128, len_x=np.inf):
...
+        """Filter the data
+
+        Parameters
+        ----------
+        method : str
```
Comment (agramfort) [inline mne/filter.py]: I would put method as second param to be able to type xxx.filter(10)

File: mne/filter.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/filter.py]: I couldn't think of a reasonable choice other than do-nothing...

File: mne/filter.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/filter.py]: I considered that, but then I worried that if we eventually add a `l_freq` param we'd end up with:
```
def filter(h_freq=None, method='savgol', l_freq=None)
```
or so, so I'm not sure what the best plan is.

Comment (Eric89GXL) [inline mne/filter.py]: I think for now I'll make it so there is no `method` param since there's only one method anyway.

**Comment by Eric89GXL** [comment]
@agramfort comments addressed. I changed the function to be `savgol_filter`, which nicely mirrors what it's called in scipy. I think this is user-friendly enough, but makes it clear it's different from e.g. `raw.filter` which uses more standard FIR/IIR filtering. You now can do `evoked.savgol_filter(10.)` like you want, too. Eventually we could expose the actual savgol params if we want, but I'll assume YAGNI.

Ready for review/merge from my end.

**Comment by agramfort** [comment]
works like a charm !

+1 for merge.

don't forget what's new

**Comment by dengemann** [comment]
Maybe worth a quick 20 lines example?

**Comment by Eric89GXL** [comment]
@dengemann done

File: mne/filter.py [hunk 3]
```diff
@@ -1389,3 +1389,68 @@ def _get_filter_length(filter_length, sfreq, min_length=128, len_x=np.inf):
...
+            data = self.data
+            axis = 1
+        elif isinstance(self, _BaseEpochs):
+            if not self.preload:
+                raise RuntimeError('data must be preloaded to filter')
+            data = self._data
+            axis = 2
+
+        if h_freq is None:
+            return
```
Comment (dengemann) [inline mne/filter.py]: Eric, what is the idea behind this return? Nothing was applied in this condition, right?

Comment (Eric89GXL) [inline mne/filter.py]: Yeah, I had it in there in case we eventually do `h_freq, l_freq` or so. I can take it out for now, though

Comment (dengemann) [inline mne/filter.py]: let's do that unless there's something useful about this empty return.

**Comment by Eric89GXL** [comment]
`is None` check removed, rebased, and updated `whats_new.py`.

**Comment by dengemann** [comment]
Please merge once builds are green

**Comment by Eric89GXL** [comment]
Done, thanks for the quick reviews. @agramfort let's see if you actually use it now :)

**Pull request merged**